Mathlib/NumberTheory/EllipticDivisibilitySequence.lean

```diff
@@ -65,26 +74,500 @@ elliptic, divisibility, sequence
+
+lemma addMulSub_swap (neg : ∀ k, W (-k) = -W k) (m n : ℤ) :
+    addMulSub W m n = - addMulSub W n m := by
+  rw [addMulSub, addMulSub, ← neg_sub, Int.neg_div, neg]; ring_nf
+
+lemma rel₃_iff₄ (m n r : ℤ) :
+    Rel₃ W m n r ↔ rel₄ W (2 * m) (2 * n) (2 * r) 0 = 0 := by
+  rw [rel₄, ← mul_zero 2, Rel₃]
+  simp_rw [addMulSub_even, add_zero, sub_zero]
+  convert sub_eq_zero.symm using 2; ring
+
+section transf
+
+variable (a b c d : ℤ)
+
+/-- The proposition that the four indices are all nonnegative and strictly decreasing. -/
+def StrictAnti₄ : Prop := 0 ≤ d ∧ d < c ∧ c < b ∧ b < a
+
+/-- The proposition that the four indices are of the same parity. -/
+def HaveSameParity₄ : Prop :=
+  a.negOnePow = b.negOnePow ∧ b.negOnePow = c.negOnePow ∧ c.negOnePow = d.negOnePow
+
+/-- The average of four indices. -/
+def avg₄ : ℤ := (a + b + c + d) / 2
+
+namespace HaveSameParity₄
+open Int Equiv
+
+variable {W a b c d} (same : HaveSameParity₄ a b c d)
+
+lemma rel₄_eq_net : rel₄ W a b c d = net W ((a - d) / 2) ((b - d) / 2) ((c - d) / 2) d := by
+  have h := @Int.two_mul_ediv_two_of_even
+  rw [net_eq_rel₄, h, h, h]; · simp_rw [sub_add_cancel]
+  all_goals simp only [← negOnePow_eq_iff, same.1, same.2.1, same.2.2]
+
+lemma even_sum : Even (a + b + c + d) := by
+  simp_rw [← negOnePow_eq_one_iff, negOnePow_add,
+    same.1, same.2.1, same.2.2, units_mul_self, one_mul, units_mul_self]
+
+lemma avg₄_add_avg₄ : avg₄ a b c d + avg₄ a b c d = a + b + c + d := by
+  rw [← two_mul]; exact Int.mul_ediv_cancel' same.even_sum.two_dvd
+
+lemma same₀₃ : a.negOnePow = d.negOnePow := by rw [same.1, same.2.1, same.2.2]
+
+protected lemma abs : HaveSameParity₄ |a| |b| |c| |d| := by
+  simpa only [HaveSameParity₄, negOnePow_abs] using same
+
+lemma perm (σ : Perm (Fin 4)) :
+    ∀ t : Fin 4 → ℤ, HaveSameParity₄ (t 0) (t 1) (t 2) (t 3) →
+      HaveSameParity₄ (t (σ 0)) (t (σ 1)) (t (σ 2)) (t (σ 3)) := by
+  have := (Perm.mclosure_swap_castSucc_succ 3).symm ▸ Submonoid.mem_top σ
+  refine Submonoid.closure_induction this ?_ (fun _ ↦ id) fun σ τ hσ hτ t same ↦ ?_
+  on_goal 2 => simp_rw [Perm.mul_apply]; exact hτ (t ∘ σ) (hσ _ same)
+  rintro _ ⟨i, rfl⟩ t ⟨h₀₁, h₁₂, h₂₃⟩; fin_cases i
+  exacts [⟨h₀₁.symm, h₀₁ ▸ h₁₂, h₂₃⟩, ⟨h₀₁ ▸ h₁₂, h₁₂.symm, h₁₂ ▸ h₂₃⟩, ⟨h₀₁, h₁₂ ▸ h₂₃, h₂₃.symm⟩]
+
+lemma six_le_of_strictAnti₄ (anti : StrictAnti₄ a b c d) : 6 ≤ a := by
+  simp_rw [HaveSameParity₄, negOnePow_eq_iff] at same
+  obtain ⟨hd, hdc, hcb, hba⟩ := anti
+  rw [← add_two_le_iff_lt_of_even_sub] at hdc hcb hba
+  · linarith
+  exacts [same.1, same.2.1, same.2.2]
+
+variable (W) in
+/-- A hybrid product formed by one factor from an `addMulSub` and one from another `addMulSub`. -/
+def addMulSub₄ (a b c d : ℤ) : R := W ((a + b).div 2) * W ((c - d).div 2)
+
+lemma addMulSub₄_mul_addMulSub₄ :
+    addMulSub₄ W a b c d * addMulSub₄ W c d a b = addMulSub W a b * addMulSub W c d := by
+  simp_rw [addMulSub₄, addMulSub]; ring
+
+lemma addMulSub_transf :
```

Comment:
I feel like `transf` really doesn't justify what these lemmas are doing.

Reply:
I added this docstring to explain what the transformation is:
```
/-! Preservation of `rel₄`, `HaveSameParity₄`, and `strictAnti₄` under the transformation
  `(a,b,c,d) ↦ (avg₄ a b c d - d, avg₄ a b c d - c, avg₄ a b c d - b, |avg₄ a b c d - a|)`. -/
```